Added animations on the first icon

ClaraChaouat opened the pull request:
Hi @sandrina-p 

## Summary 

I added "animations" on the first icon. (It's probably not the DRYest code you've ever seen ;)).
By clicking on the first icon, it disappears and then another icon (a dot) displays.
Then, by clicking the dot, the dot disappears and the previous icon displays again.
I guess It's a first "animation" step. I know that the way icons display and disappear is not what we want. I think it has something to do with a "toggle" thing If I'm correct. I still need to investigate.  

 <!-- Your major learnings and WOW moments during this challenge --> ## Struggles

The idea was to build this animation using vanilla javascript. 
At the beginning, I tried to find an "easy" solution on stackoverflow (I'm being honest ;)) but I couldn't find something that fitted (maybe it's because I still have to improve my "how to type my question" skills). So, following the advice you gave me on the very first day when I asked you what do we do when super stuck? You said: "start with very simple" things. This is what I did. I started by looking for a way to do an "onclick" event on a href. Then I broke into mini steps what I wanted to achieve:
- Click -> the calendar icon disappears -> the dot icon displays -> click again the dot icon disappears -> the calendar icon displays back.

What I learnt: how to use the class .active , how to add/remove a class with an onclick event.
Regarding git/github: I got that we don't merge a branch until the work on that branch is not done.
Struggles : I struggled when trying to understand where to start from.

 

 <!-- What did you not understand or struggled the most to figure out --> ## References 
I still need to understand how we call the way icons are supposed to be displayed/ disappear (toggling thing?) and how to achieve that.
I'm sure this code could be improved with a dryer version of it.

<!-- Resources links you found useful while working on this -->
https://codepen.io/tylerama/pen/nLFHt
https://clubmate.fi/detect-click-with-pure-javascript

Thank you. I'm waiting for your feedback :) .

Clara.

sandrina-p commented on animations.js:
```diff
@@ -30,4 +24,4 @@ let clickHandlerCircle = function () {
 }
 
-circleIcon.addEventListener('click', clickHandlerCircle)
+circleIcon.addEventListener('click', clickHandlerCircle)
```
Nice work doing the circle click event. A tip:

You can use `circleIcon.classList.toggle('active')` if not existing, it set active class, otherwise remove the it.

For now, it's totally okay to do 4 event listeners, one for each icon. Later we can DRY it too.

sandrina-p commented on index.html:
```diff
@@ -10,15 +10,15 @@
                     <li class="calendare">
                         <a id="calendar" class=" calendar-icon active" href="#"><img
-                                src="./calendar-plus-solid.svg"></a>
+                                src="./calendar-plus-solid.svg"></a><a id="circle" class="circle-icon" href="#"><img
```
I believe you didn't stage+commit+push the `circle-solid` image, because on my computer I cannot see it.

Nevertheless, why don't you try the following:

- Instead of an image for the dot, create an empty div and use CSS border + bacgrkound properties. By default, it has opacity 0. When the link is active, it sets opacity to 1.  
- The icon itself does not need to disappear (display: none). By doing that, you remove it from the layout and it breaks the animations. So, it's better to also use the opacity tricky.

Another element missing is the link text. To keep things easy, the opacity trick can also be used.

I hope this helps you!

sandrina-p commented on style.css:
```diff
@@ -33,3 +33,30 @@ a {
+	display: none;
+}
+
+.circle-icon.active {
```
Instead of creating a class for each icon, you could add a common class to every icon for example `link`, and then reuse it. :)

sandrina-p replied:
Based on this CSS I think you are not familiar with CSS specificity and selectors.

Here's an interactive tutorial (aka a CSS game) to help you learn them: https://flukeout.github.io/

sandrina-p commented:
Here's a suggestion to organize your files inside the project. You could move all the images to a `images/` folder, to keep the project root cleaner :)

ClaraChaouat commented:
Hi @sandrina-p,

You'll see that there is a space between the dot and the top of the page I'm not able to understand where does that come from ?

ClaraChaouat commented:
Hi @sandrina-p

## Summary 
I almost added all the suggestions you gave me.
I added "motion" on the calendar after watching the tutorial you sent me (It's nice to see how you can create very nice things with very simple and efficient code). I still have to add motion on the link text.
## Learnings
 - How to use css transform properties.
 - Using display none is not a good practice and how to play with opacity instead.
## Struggles
I have two questions:
- I'm wondering why when I click on the calendar icon it makes the other icons move too?
- If you have a look at line 68 in the style.css file, I had to set a negative margin to remove the space between the dot and the text. I could figured out where does that space come from. Any clue ?
## References 
https://www.youtube.com/watch?v=Nloq6uzF8RQ
https://www.youtube.com/watch?v=zHUpx90NerM
https://www.youtube.com/watch?v=ypwVbSAAJ68

Thanks for your feedback :).

sandrina-p reviewed (commented):
Hi Clara, I've decided to review this with a little video recording, I've sent you the link on Discord. Good luck!

ClaraChaouat commented:
Hi @sandrina-p 

I hope you're doing well.

## Summary 
I added scale and translate animations on the first navbar item.

## Learning
- How/when to use the absolute position (I never had the occasion to use it before, now this property makes more sense to me).
- How to use the chrome tool to "play" with the animations.
- An anchor tag can "nest" other tags.
- Building a dot with css only.
- How to style a global selector except one.
- css variables (I still need to investigate on that topic).

 ## Struggles
 - I still have to work on the animations (the "flashing/bubbling" effect right after clicking and the way the calendar icon disappears in diagonal, I haven't investigated yet).
 - I wrote some questions on notion about css.
 
 ## References 
Sorry I forgot to let the pages Which helped me open.

Thank you :) .

sandrina-p commented on animations.js:
```diff
@@ -1,33 +1,9 @@
-//console.log('Click just happened')
-//}
+// First Icon event handling
+elMenuItem.addEventListener('click', () => {
```
Now, to have to add an event listener to each one of the menu items, you could hardcode it and copy-paste, or you could DRY it.

Hint: look for ways to do loops in JS. [`forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) can be a solution.

sandrina-p commented on style.css:
```diff
@@ -23,13 +27,68 @@ ul {
 }
 
-a {
+a:not(.item-link) {
```
When I did this, was for a  debug/demonstration. In the final solution, I'd remove the `a` selector completely

sandrina-p commented on style.css:
```diff
@@ -23,13 +27,68 @@ ul {
+}
+
+.item-link.active .item-icon {
+	/* Select the element with the className "item-link" and "active" and "item-icon"?*/
```
No, Select the `item-icon` when the `.item-link` has the `.active` class. I highly recommend you to learn and practice more about CSS specificity :) 

This is a nice way to learn while playing a game. https://flukeout.github.io/

ClaraChaouat replied:
Coud we also use this synthax ?

 .item-link:active + .item-icon {
}

ClaraChaouat replied:
Or something like this ?
.item-link:active, .item-icon {
}

sandrina-p replied:
Yes that syntax is also valid but it means different type of selections.

ClaraChaouat replied:
Thanks :).
I was reading about css specificity and I was wondering If the above one could work.

ClaraChaouat replied:
What do you mean by "it means different type of selections"?

sandrina-p replied:
- `.a .b` is [descedant selector](https://developer.mozilla.org/en-US/docs/Web/CSS/Descendant_combinator)
- `.a + .b` is [adjacent sibling combinator](https://developer.mozilla.org/en-US/docs/Web/CSS/Adjacent_sibling_combinator) 
- `.a:active` is [a pseudo class](https://developer.mozilla.org/en-US/docs/Web/CSS/:active)

More [CSS selectors](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Selectors)

ClaraChaouat replied:
Thanks. I added the link to my favorites!

sandrina-p reviewed (commented):
This is starting to look nice! Left some comments to guide you further ;)

sandrina-p commented on style.css:
```diff
@@ -23,13 +27,68 @@ ul {
+	background-color: var(--color);
+	opacity: 0;
+	transform:scale(0);
+	transition:all 500ms;
```
Avoid using `all`. This animates everything, even things that you don't need and can make a transition slower. Always be specific about the things you really want to animate.

sandrina-p commented on style.css:
```diff
@@ -23,13 +27,68 @@ ul {
+.item-link.active .item-dot {
+	opacity: 1;
+	transform: translateY(-5px);
+	transform:scale(1); /*1 is the original size*/
```
In CSS you can only use a property once. When repeated, it will apply only the last one.

In this case, the `translateY` was ignored. So, you need to merge both together:

```
transform: translateY(-5px) scale(1);
```

ClaraChaouat replied:
Oh ans apparently the order matters. The properties will be applied from right to left. Interesting...

sandrina-p replied:
does the word matter for transform? I didn't know that. Can you share where you learned that?

ClaraChaouat replied:
Here is what I found about this topic:

https://stackoverflow.com/questions/10765755/how-to-apply-multiple-transforms-in-css

sandrina-p replied:
Oh, that's true!! I didn't remember that at all, nice find!

ClaraChaouat commented:
Hi @sandrina-p 

Here are the small steps I did.

## Summary 
The idea was to understand how the "explosion/bubble" effect was built and start coding it.
## Learnings
 - I started by coding a circle with no background and with a border which has the same color as the icons.
 - I also tried to understand the timing of all the animations. So I added some transition delays to some animations.
 - By trying to reduce the width of the circle border, I learnt that the thinest border size is 1px and If we want to set a "thiner" border with, we have to play with the alpha value (https://stackoverflow.com/questions/13891177/css-border-less-than-1px)

 ## Struggle
- I struggled for a long time trying to understand why the border was not showing up. I finally debugged myself (It was a "WOW" moment) when I realised that I didn't use the right css property. For some reason, I wrote _border-color: 1px solid whatever the color_ instead of _border: 1px solid whatever the color_.

- Question: when you adviced me to play with the width for the "circle explosion" animation. Did you mean the width of the border or the one of the div containing the "circle explosion"?

 ## References 
I will probably get inspiration from that link to keep on working on the circle explosion animation: 

https://www.w3schools.com/cssref/tryit.asp?filename=trycss_anim_border

ClaraChaouat commented:
Hi @sandrina-p ,

## Summary 
- I "polished" the explosion animation,
- I added a clip-path to the icon animation. I find mine less "smooth" than the one from Dribbble,
- I changed the item-text font color.

## Learnings
- Creating an explosion effect by playing with the margin width.
-Top css property: as I think I had never used the property position absolute before, I didn’t know about the top property, which I found handy.
- Clip-path: I didn't know we called that effect like this and how to use it.

 ## Struggles
- I didn't really struggle this time :).
- Clip-path transition: I still think It could be improved. What do you think? If yes, could you please give me a hint? (but not the answer;)).
- Gif preview: as I'm not on Mac, I was not able to use the preview trick you gave me but It "inspired" me. I found a website where you can either reduce the GIF speed or extract frames from the it. 

## References 
- Gif speed reducer/extractor: https://ezgif.com/speed
-  Clip-path: https://bennettfeely.com/clippy/  I found this link very useful to understand how to set the coordinate system when using clip-path.

Thanks for your feedback :) .

sandrina-p reviewed (commented):
You're almost there! The animations timing still need some tweaks, they are too fast. Try to fix each one at a time. Another thing that will help you is `transition-timing-function`.

Good luck!

sandrina-p commented on README.md:
```diff
@@ -1,2 +1,27 @@
+- Type git clone and then paste the URL you copied earlier,
+- Press enter to create a local clone.
+
+## Demo
```
Uh demo is a nice idea! Do you know how to publish a website from GitHub? There are 2 popular ways:
- Github Pages
- Netlify

Personally I prefer [Netlify](https://www.netlify.com/). It's more [friendly and powerful](https://www.aleksandrhovhannisyan.com/blog/github-pages-vs-netlify/) and it supports deploys from PR. Give it a try!

ClaraChaouat replied:
I planned to insert the GIF in the README.md but I can do both If you think that I could be a good exercice for me.

sandrina-p replied:
Yes, try to integrate Netlify on this github repo. When integrated, a new deploy should run on this PR after each commit pushed.

ClaraChaouat commented on style.css:
```diff
@@ -23,13 +36,94 @@ ul {
+	margin-bottom: 5px;
+	opacity: 1;
+	transform: translateY(0px);
+	clip-path: polygon(0 0, 100% 0, 100 0);
```
I did yes :)!

ClaraChaouat commented on README.md:
```diff
@@ -1,2 +1,27 @@
+This challenge helped to realize how we can build sophisticated things with simple code.
+
+## Installation
+Clone this repository to get a version in your local environment.
```
Hahah yeah, I tried to be in my shoes months ago when I started with git :)

ClaraChaouat commented:
Hi @sandrina-p,

I hope you're doing well.
I think we are almost there, maybe just one more detail in the js file to fix. 


## Summary 
### CSS
- Fixed the clip-path issue (from square to triangle. One pair of coordinates was missing),
- Adjusted the animations transition timing.
### HTML
- Added HTML for the other icons.

### JS
- "DRYed" the animations code. But, If we pay attention to the animation; on the original version of the toolbar, once we click again to the same item (active class -> non active class), it activates another item at the same time. Do you think I should fix that ?

### GIF
- Created a GIF (first one but not the least).

### Git/Git hub
- I think I did my first git pull. I edited the README file directly from github, I think that is why I had to git pull when getting back to vs code to commit and put changes on the js, HTML and css files.

## Learnings
- How to create a GIF (WOW moment ;)). I learnt that you can hide the cursor (I was wondering how the guy from dribbble did). Also, I learnt that in order an animation never stops, It has to "loop forever" (I was also wondering how the guy from dribbble did that). 
- How to use transition-timing function.
 
 ## References 
https://www.youtube.com/watch?v=jHpk1JS44yE
https://www.screentogif.com/ (I created the GIF with this one)
https://developer.mozilla.org/fr/docs/Web/CSS/transition-timing-function

sandrina-p commented on README.md:
```diff
@@ -1,2 +1,21 @@
+- Vanilla JS
+
+## Credit
+This toolbar icons animation is originally designed by Cuberto (https://dribbble.com/shots/5605168-Toolbar-icons-animation)
```
You can create links with markdown.

```
[designed by Cuberto]https://dribbble.com/shots/5605168-Toolbar-icons-animation)
```

sandrina-p commented on animations.js:
```diff
@@ -1,33 +1,13 @@
+  array
+) {
+  element.addEventListener('click', () => {
+    element.classList.toggle('active')
```
Well done, this is perfect! Now it's just missing the logic to remove the `active` from the remaining items.

Hint:
Option A: inside the event listener, before the toggle, do another loop to remove the class from all the items. 

Option B (more efficient, but more complex): When clicking in an item using the `index` to store the latest clicked item at `activeIndex`. Use that global variable to decide which item.active should be removed in new clicks.

sandrina-p replied:
I see you opt-in for the Option A. Here's the solution for Option B:

```js
let elMenuItems = document.getElementsByClassName("item-link");

// -1 by default means there's no current element
let activeIndex = -1;

Array.from(elMenuItems).forEach(function (element, index) {
  element.addEventListener("click", () => {
    // If there's an active menu, and it's different from the current
    if (activeIndex > -1 && activeIndex !== index) {
      // ...remove .active from it
      elMenuItems[activeIndex].classList.remove("active");
    }

    // Toggle active to the clicked element
    element.classList.toggle("active");
    activeIndex = index;
  });
});

```

sandrina-p commented on style.css:
```diff
@@ -23,13 +36,88 @@ ul {
+
+.item-link.active .item-icon {
+	opacity: 0;
+	clip-path: polygon(0 0, 100% 0, 100% 100%);
```
The clip-path animation is still broken. You can use firefox, their dev tools are awesome to play around with clip-path: (see that little icon in the screenshot).

Tip: Disable the JS click event and transitions while playing with devtools polygon edition, to avoid weird conflicts/delays.

![image](https://user-images.githubusercontent.com/14869087/106684719-e8de3380-65be-11eb-9cf0-73f54d0c2f7e.png)

ClaraChaouat commented:
Hi @sandrina-p ,

I think we're good.
I'm waiting for your "go" to merge it.
The firefox mozilla dev tool was a great tip!

sandrina-p reviewed (approved):
Well done! 🎉

sandrina-p commented on README.md:
```diff
@@ -1,2 +1,26 @@
+## Technologies used
+- HTML
+- CSS
+- Vanilla JS
```
If you want, you could create a section `## Learnings` with a summary of your major learnings with this challenge ;)
